### PR TITLE
Reimplement IPSet.Ranges with a (somewhat) easier to follow algorithm.

### LINKED ISSUE #125: IPSet.Ranges sometimes generates invalid IPRanges
Found while implementing other functionality and writing tests. PR at #124 adds an assert that trips on the existing single_ips test.

```
--- FAIL: TestIPSet (0.00s)
    --- FAIL: TestIPSet/single_ips (0.00s)
        ipset.go:162: post-sort:
        ipset.go:142:  {  10.0.0.0        ✅
        ipset.go:144:   } 10.0.0.0        ✅
        ipset.go:142:  {  10.0.0.1        ✅
        ipset.go:144:   } 10.0.0.1        ✅
        ipset.go:142:  {  10.0.0.2        ✅
        ipset.go:144:   } 10.0.0.2        ✅
        ipset.go:142:  {  10.0.0.3        ✅
        ipset.go:144:   } 10.0.0.3        ✅
        ipset.go:142:  {  10.0.0.4        ❌
        ipset.go:142:  {  10.0.0.4        ✅
        ipset.go:144:   } 10.0.0.4        ❌
        ipset.go:144:   } 10.0.0.4        ✅
        ipset.go:164: merging...
        ipset.go:183: at[0] ({ip:{addr:{hi:0 lo:281470849515520} z:0xc000128060} want:true start:true}), add=1, remove=0
        ipset.go:183: at[1] ({ip:{addr:{hi:0 lo:281470849515520} z:0xc000128060} want:true start:false}), add=0, remove=0
        ipset.go:183: at[2] ({ip:{addr:{hi:0 lo:281470849515521} z:0xc000128060} want:true start:true}), add=1, remove=0
        ipset.go:183: at[3] ({ip:{addr:{hi:0 lo:281470849515521} z:0xc000128060} want:true start:false}), add=0, remove=0
        ipset.go:183: at[4] ({ip:{addr:{hi:0 lo:281470849515522} z:0xc000128060} want:true start:true}), add=1, remove=0
        ipset.go:183: at[5] ({ip:{addr:{hi:0 lo:281470849515522} z:0xc000128060} want:true start:false}), add=0, remove=0
        ipset.go:183: at[6] ({ip:{addr:{hi:0 lo:281470849515523} z:0xc000128060} want:true start:true}), add=1, remove=0
        ipset.go:183: at[7] ({ip:{addr:{hi:0 lo:281470849515523} z:0xc000128060} want:true start:false}), add=0, remove=0
        ipset.go:183: at[8] ({ip:{addr:{hi:0 lo:281470849515524} z:0xc000128060} want:false start:true}), add=0, remove=1
        ipset.go:183: at[9] ({ip:{addr:{hi:0 lo:281470849515524} z:0xc000128060} want:true start:true}), add=1, remove=1
        ipset.go:183: at[10] ({ip:{addr:{hi:0 lo:281470849515524} z:0xc000128060} want:false start:false}), add=1, remove=0
        ipset.go:183: at[11] ({ip:{addr:{hi:0 lo:281470849515524} z:0xc000128060} want:true start:false}), add=0, remove=0
        ipset.go:225: post-merge:
        ipset.go:142:  {  10.0.0.0        ✅
        ipset.go:144:   } 10.0.0.3        ✅
        ipset.go:142:  {  10.0.0.5        ✅
        ipset.go:144:   } 10.0.0.4        ✅
panic: internal error; end before start [recovered]
	panic: internal error; end before start
```

The sort looks reasonable, but post-merge we end up with the invalid range 10.0.0.5-10.0.0.4.

cc @bradfitz

### PR DESCRIPTION
**Pull request by danderson**
```
    The main goal was to fix #125 by providing a less tricky implementation
    of Ranges as a comparison baseline... But this implementation turns out
    to also be a bit faster and more memory efficient.
    
    Fixes #125.
    
    name         old time/op    new time/op    delta
    IPSetFuzz-8    16.3µs ± 3%    14.1µs ± 1%  -12.99%  (p=0.008 n=5+5)
    
    name         old alloc/op   new alloc/op   delta
    IPSetFuzz-8    2.60kB ± 0%    1.95kB ± 0%  -25.24%  (p=0.008 n=5+5)
    
    name         old allocs/op  new allocs/op  delta
    IPSetFuzz-8      33.0 ± 0%      30.0 ± 0%   -9.09%  (p=0.008 n=5+5)
    
    Signed-off-by: David Anderson <dave@natulte.net>
```

### CODE REVIEW

**Comment by danderson** [comment]
Ready for review @bradfitz @mdlayher. Tests pass at all commits (I only add the previously-failing test case after the major commit that fixes the bug), so can be rebase-and-merged.

**Review by mdlayher — APPROVED** [review]
LGTM, and I appreciate the ASCII diagrams!

File: ipset.go [hunk 1]
```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
-			*depth++
-		} else {
-			*depth--
-		}
+	// in and out are sorted in ascending range order, and have no
+	// overlaps within each other. We can run a merge of the two lists
+	// in one pass.
+
+	ret := make([]IPRange, 0, len(in))
+	for len(in) > 0 && len(out) > 0 {
```
Comment (bradfitz) [inline ipset.go:131]: Maybe give in[0] and out[0] names to eliminate some of the syntactical stutter below?
 
```
rin, rout := in[0], out[0]
```

or `inr, outr`

File: ipset.go [hunk 2]
```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
-					want:  true,
-					start: true,
-				})
+
+		switch {
+		case !out[0].Valid() || !in[0].Valid():
+			// mergeIPRanges should have prevented invalid ranges from
+			// sneaking in.
+			panic("invalid IPRanges during Ranges merge")
+		case out[0].To.Less(in[0].From):
```
Comment (bradfitz) [inline ipset.go:141]: seems like most of these cases could/should be unexported methods on IPRange would let you move docs to the methods, and let them be tested easily separately too.

```go
   case outr.entirelyBefore(inr):
```

(or out[0] if you don't like names for them)

File: ipset.go [hunk 3]
```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+		ret = append(ret, in...)
+		if debug {
+			debugf("ret=%v", ret)
+		}
 	}
 
-	if len(want)%2 == 1 {
-		panic("internal error; odd number")
-	}
+	// TODO: this could update s.in and s.out opportunistically, so
```
Comment (bradfitz) [inline ipset.go:227]: Not until https://github.com/inetaf/netaddr/issues/110 is figured out.

File: netaddr.go [hunk 4]
```diff
@@ -1245,6 +1248,68 @@ func (r IPRange) contains(addr IP) bool {
 	return r.From.Compare(addr) <= 0 && r.To.Compare(addr) >= 0
 }
 
+func (r IPRange) less(other IPRange) bool {
```
Comment (bradfitz) [inline netaddr.go:1251]: ah, there is one method. worth some docs about what "less" means on an IPRange.

File: netaddr.go [hunk 5]
```diff
@@ -1245,6 +1248,68 @@ func (r IPRange) contains(addr IP) bool {
...
+	if cmp := r.From.Compare(other.From); cmp != 0 {
+		return cmp < 0
+	}
+	return r.To.Less(other.To)
+}
+
+// mergeIPRanges returns the minimum and sorted set of IP ranges that
+// cover r.
+func mergeIPRanges(r []IPRange) (out []IPRange, valid bool) {
```
Comment (bradfitz) [inline netaddr.go:1260]: for a slice I'd use `rr`, `rs`, or `rv` , then you can use `r` as your iteration variable on line 1273

File: ipset.go [hunk 6]
```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+			//
+			//    in         out
+			// f------t   f-------t
+			ret = append(ret, in[0])
+			in = in[1:]
+			if debug {
+				debugf("in before out; append in")
+				debugf("ret=%v", ret)
+			}
+		case out[0].From.lessOrEq(in[0].From) && in[0].To.lessOrEq(out[0].To):
```
Comment (bradfitz) [inline ipset.go:161]: case foo.supersetOf(bar):

(if you like "superset"?  it's not ambiguous like "covers" which can be partially or fully covered)

**Review by bradfitz — COMMENTED** [review]
Nice!

**Comment by danderson** [comment]
Looks much better with the symbolic names for each case, thanks. Applied all feedback, going to submit and we can bikeshed on the exact names of helpers in a followup (I'm not _super_ happy with some of the subtler ones).

**Pull request merged**